blues/sparrow-reference-web-app
# feat: multiple gateways

# Problem Context

Currently the reference app takes a Device UID and uses that as the chosen gateway for the project.

As an extension, we should also be able to display multiple gateways and their associated sensors, by specifying the gateways as a comma-separated list of device UIDs in the config, or

UX - for multiple gateways. Carousel.
Update page templates: iterate over gateways
Update DataProvider interface to retrieve all the gateways
Retrieve gateways via comma separated list in the env var.

## Changes

The NotehubDataProvider can now handle either a single gateway or multiple gateways being passed in a comma separate list via its environment variable file `.env.local`.
Added carousel on the frontend to handle multiple gateways from a user's perspective, and users can click arrows to paginate through all gateways present.
Added a new property to the `Gateway` model named `sensorList` in order to more easily group sensors underneath their respective gateways for display on the dashboard and to make the carousel work. Refactored the `SensorCard` component to be a child of the `GatewayCard` and pass the `GatewayCard` both gateway and sensor data as one object to display as appropriate.

## Screenshot (if applicable)

https://user-images.githubusercontent.com/20400845/155809747-5fe29b0b-56be-44cc-a466-fbcdb7ac6760.mov

## Testing

Unit / integration / e2e tests?

Unit tests added to. E2E tests updated to account for multiple gateways.

Steps to test manually?

1. Add a comma separated list of device IDs to your `.env.local` file. I used the a-team Sparrow project as we have the most Sparrow devices running simultaneously there (`HUB_DEVICE_UID=dev:89153172025c,dev:5c0272356845,dev:5c027231167f,dev:5c0272356877,dev:5c0272311675`)
2. Go to localhost:4000/
3. Observe the homepage which should display just one gateway and its respective sensors.
4. Click the buttons at the left or right to scroll through the other gateways and see their sensors as well.

Any other sorts of testing notes to include?

One e2e is failing due to the fact that it's expecting to have more than one gateway present, however altering the GH environment variables causes other PR branches not yet ready to handle multiple gateways to fail their own tests.

I ran the e2es locally against my branch and had a clean sweep. Here's a screenshot for proof.

<img width="757" alt="Screen Shot 2022-03-01 at 4 27 50 PM" src="https://user-images.githubusercontent.com/20400845/156252288-21f29617-43a8-4724-ba69-60eaee29ff45.png">


## Any other related PRs

n/a

## Ticket(s)

https://www.pivotaltracker.com/story/show/181116144
